core/src/WsTrust/MexParser.cs

```diff
@@ -0,0 +1,280 @@
+    }
+
+    internal class WsTrustAddress
+    {
+        public Uri Uri { get; set; }
+
+        public WsTrustVersion Version { get; set; }
+    }
+
+    internal class MexPolicy
+    {
+        public WsTrustVersion Version { get; set; }
+
+        public string Id { get; set; }
+
+        public UserAuthType AuthType { get; set; }
+
+        public Uri Url { get; set; }
+    }
+
+    internal static class MexParser
+    {
+        private const string WsTrustSoapTransport = "http://schemas.xmlsoap.org/soap/http";
+
+        public static async Task<WsTrustAddress> FetchWsTrustAddressFromMexAsync(string federationMetadataUrl, UserAuthType userAuthType, RequestContext requestContext)
+        // This method would return null if the wanted policy is not found in mexDocument. It could potentially throw XmlException or http-relevant exceptions too.
+        {
+            XDocument mexDocument = await FetchMexAsync(federationMetadataUrl, requestContext).ConfigureAwait(false);
+            return ExtractWsTrustAddressFromMex(mexDocument, userAuthType, requestContext);
+        }
+
+        internal static async Task<XDocument> FetchMexAsync(string federationMetadataUrl, RequestContext requestContext)
+        {
+            var httpResponse = await HttpRequest.SendGet( // It would potentially throw MsalServiceException
+                new UriBuilder(federationMetadataUrl).Uri, null, requestContext).ConfigureAwait(false);
+            if (httpResponse.StatusCode != System.Net.HttpStatusCode.OK) {
+                throw new Client.MsalException(
+                    MsalError.AccessingWsMetadataExchangeFailed,
+                    string.Format(CultureInfo.CurrentCulture,
+                        "Response status code does not indicate success: {0} ({1}).",
+                        (int)httpResponse.StatusCode, httpResponse.StatusCode));
+            }
+            return XDocument.Parse(httpResponse.Body, LoadOptions.None);
+        }
+
+        internal static WsTrustAddress ExtractWsTrustAddressFromMex(XDocument mexDocument, UserAuthType userAuthType, RequestContext requestContext)
+        // This method would return null if the wanted policy is not found in mexDocument. It could potentially throw XmlException.
+        {
+            WsTrustAddress address = null;
+            MexPolicy policy = null;
+
+            Dictionary<string, MexPolicy> policies = ReadPolicies(mexDocument);
+            Dictionary<string, MexPolicy> bindings = ReadPolicyBindings(mexDocument, policies);
+            SetPolicyEndpointAddresses(mexDocument, bindings);
+            Random random = new Random();
+            //try ws-trust 1.3 first
+            policy = policies.Values.Where(p => p.Url != null && p.AuthType == userAuthType && p.Version == WsTrustVersion.WsTrust13).OrderBy(p => random.Next()).FirstOrDefault() ??
+                        policies.Values.Where(p => p.Url != null && p.AuthType == userAuthType).OrderBy(p => random.Next()).FirstOrDefault();
```

Comment:
This conversation was talking about line 103. The following response is derived from my earlier reply from Github:

Now that github displays [the old code](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118/files#diff-96025f3ff8136252b94d8c9a98c11b26L114) and new code together, we can see this part is exactly the same as before. Let me know if you still want to wrap this into its own method.

---
In reply to: [201758554](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r201758554) [](ancestors = 201758554)

Reply:
you are right it is about line 103. and even though things were like before it doesn't make it easier to read. imo it would better to make a method which name indicates what's going on. that would also alleviate the need for the comment in the code... it's anyways not telling the full story, unless I misread the code (which is very likely).

---
In reply to: [207083470](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r207083470) [](ancestors = 207083470,201758554)

Reply:
On a closer look, your doubt was on "randomness is weird; why is it needed". And the reason for the randomness was undocumented when [that line was added at 4 years ago](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blame/v3.19.8/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/WsTrust/MexParser.cs#L107). Right now I can only guess that, perhaps that developer wanted to have an more unbiased representation across multiple runs, in the case when multiple policies were available. I can document my such guess in a new helper function `SelectPolicy(...)`. Or, if you prefer, we can simply remove those randomness thing and always use the first available policy. In the latter case, do we still need a standalone helper function for that? Please let me know.

---
In reply to: [207387634](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r207387634) [](ancestors = 207387634,207083470,201758554)